ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/MyAppPurchaseLogic.kt

```diff
@@ -0,0 +1,118 @@
+package com.revenuecat.purchases.ui.revenuecatui
+
+import android.app.Activity
+import com.android.billingclient.api.Purchase
+import com.revenuecat.purchases.CustomerInfo
+import com.revenuecat.purchases.Package
+import com.revenuecat.purchases.PresentedOfferingContext
+import com.revenuecat.purchases.ProductType
+import com.revenuecat.purchases.PurchasesError
+import com.revenuecat.purchases.models.GoogleReplacementMode
+import com.revenuecat.purchases.models.PurchaseState
+import com.revenuecat.purchases.models.PurchaseType
+import com.revenuecat.purchases.models.StoreTransaction
+import org.json.JSONObject
+
+/**
+ * Code for handling in-app purchases and restorations directly by the application rather than by RevenueCat.
+ * These methods are called by a RevenueCat Paywall only when `Purchases.purchasesAreCompletedBy` is `MY_APP`.
+ *
+ * @property performPurchase Performs an in-app purchase.
+ * @property performRestore Perform a purchase restore for the given customer.
+ */
+class MyAppPurchaseLogic(
+    val performPurchase: suspend ((Activity, Package) -> MyAppPurchaseResult),
+    val performRestore: suspend ((CustomerInfo) -> MyAppRestoreResult),
```

Comment:
I was wondering, until now, we've been using lambda functions instead of coroutines for things like this (and then wrap those into async APIs in some cases. Is there a reason we wanted to go coroutine-first here? 

FWIW, we've discussed a bit about going with coroutines moving forward, but we should make sure that we keep compatibility for people not using coroutines and we keep a similar API.

IMO, for now, we could have this be an interface with these 2 as functions, that need to be implemented by the developer. This is also the approach we follow with the `PaywallListener`.

Reply:
Just double checking, with this: 
>Right, in this case I'm actually thinking that to make it simpler on developers, we want to get the response on a callback itself. 
  
are you referring to the `completion` callback in your suggested code block? 

I was actually going to suggest the option of making them blocking with return values (as I believe the BillingClient's `launchBillingFlow()` is blocking as well?): 
```diff
interface MyAppPurchaseLogic {
-    fun performPurchase(activity: Activity, rcPackage: Package, completion: (MyAppPurchaseResult) -> Unit)
+    fun performPurchase(activity: Activity, rcPackage: Package): MyAppPurchaseResult
-    fun performRestore(completion: (MyAppRestoreResult) -> Unit)
+    fun performRestore(): MyAppRestoreResult
}
``` 
  
If **not** doing that ☝️ is what you meant, I believe you 😄 You have more context around these APIs and how they're typically used.

Reply:
Actually `launchBillingFlow` is asynchronous. As in, it doesn't block the thread, but launches a new window/activity on top to display the purchase flow.

So yeah, we shouldn't make them blocking, and providing a completion block/asynchronous function would help in that sense.

Honestly, it would look much nicer with the coroutine approach, but I'm not sure of suddenly changing the API style for this... I could be convinced otherwise though.

Reply:
>Actually `launchBillingFlow` is asynchronous.
  
You're right! 

>Honestly, it would look much nicer with the coroutine approach, but I'm not sure of suddenly changing the API style for this... I could be convinced otherwise though.

Well, one could argue that `PurchaseLogic` is a different beast from `PaywallListener`. The former implements long-running operations, while the latter is just notifying you of events that happened. Following that argument it is not super strange that they have different APIs. 

If we're doing `suspend` functions, how are we going to do that in `PaywallActivity`, which could be used from Java? I guess we could use a [`lifecycleScope`](https://developer.android.com/topic/libraries/architecture/coroutines#lifecyclescope) to call the `suspend` functions, and turn them into callbacks for Java. (`lifecycleScope` is part of `androidx.lifecycle:lifecycle-runtime-ktx`, which we are already (transitively) depending on.)

Not sure where I stand yet haha. I can agree to both I think. `suspend` functions are definitely an elegant solution.

Reply:
Thank you for this vigorous discussion! I will look for direction, whatever the conclusion is.

Reply:
Sorry I missed this conversation but wanted to chime in as well by expressing my opinion, which I believe is the same as yours.

I'd vote for not using coroutines as well and have a regular callbacks because it gives more flexibility, and if someone wants to use coroutines themselves because their purchasing code uses coroutines they still can by wrapping it in a coroutine. If we make it suspend, we are forcing everyone to use coroutines, and some people might not be using those, specially if we take into account that launchBillingFlow doesn't have a suspended version in the billing-ktx library.